builders/ant.go

```diff
@@ -0,0 +1,222 @@
+
+// Initialize collects metadata on Java and SBT binaries
+func (builder *AntBuilder) Initialize() error {
+	antLogger.Debugf("Initializing Ant builder...")
+
+	// Set Java context variables
+	javaCmd, javaVersion, err := which("-version", os.Getenv("JAVA_BINARY"), "java")
+	if err != nil {
+		antLogger.Warningf("Could not find Java binary (try setting $JAVA_BINARY): %s", err.Error())
+	}
+	builder.JavaCmd = javaCmd
+	builder.JavaVersion = javaVersion
+
+	// Set Ant context variables
+	antCmd, antVersionOut, err := which("-version", os.Getenv("ANT_BINARY"), "ant")
+	if err != nil {
+		return fmt.Errorf("could not find Ant binary (try setting $ANT_BINARY): %s", err.Error())
+	}
+
+	builder.AntCmd = antCmd
+	antVersionMatchRe := regexp.MustCompile(`version ([0-9]+\.[0-9]+.\w+)`)
+	match := antVersionMatchRe.FindStringSubmatch(antVersionOut)
+
+	if len(match) == 2 {
+		builder.AntVersion = match[1]
+	}
+
+	antLogger.Debugf("Initialized Ant builder: %#v", builder)
+	return nil
+}
+
+// Build is currently not implemented
+func (builder *AntBuilder) Build(m module.Module, force bool) error {
+	return errors.New("Ant builder is not implemented")
+}
+
+// Analyze resolves a lib directory and parses the jars inside
+func (builder *AntBuilder) Analyze(m module.Module, allowUnresolved bool) ([]module.Dependency, error) {
+	antLogger.Debugf("Running Ant analysis: %#v %#v in %s", m, allowUnresolved, m.Dir)
+
+	options := m.Context.(map[string]interface{})
+
+	libdir := options["libdir"].(string)
+	if libdir == "" {
+		libdir = "lib"
+	}
+
+	libDirExists, err := hasFile(m.Dir, libdir)
+	if !libDirExists || err != nil {
+		return nil, errors.New("unable to resolve library directory, try specifying it using the `modules.options.libdir` property in `fossa.yml`")
+	}
+
+	jarFilePaths, err := doublestar.Glob(filepath.Join(libdir, "*.jar"))
+	if err != nil {
+		return nil, err
+	}
+
+	antLogger.Debugf("Running Ant analysis: %#v", jarFilePaths)
+
+	var dependencies []module.Dependency
+
+	// traverse through libdir and and resolve jars
+	for _, jarFilePath := range jarFilePaths {
+		locator, err := locatorFromJar(jarFilePath)
+		if err == nil {
+			dependencies = append(dependencies, module.Dependency{
+				Locator: locator,
+			})
+		} else {
+			antLogger.Warningf("unable to resolve Jar: %s", jarFilePath)
+		}
+	}
+
+	return dependencies, nil
+}
+
+// locatorFromJar resolves a locator from
+func locatorFromJar(path string) (module.Locator, error) {
+	antLogger.Debugf("processing locator from Jar: %s", path)
+
+	info, err := jargo.GetJarInfo(path)
+	if err == nil {
+		// first, attempt to resolve a pomfile from the META-INF directory
+		var pomFilePath string
+		for _, file := range info.Files {
+			if strings.HasPrefix(file, "META-INF") && strings.HasSuffix(file, "pom.xml") && (pomFilePath == "" || len(pomFilePath) > len(file)) {
+				pomFilePath = file
+			}
+		}
+
+		if pomFilePath != "" {
+			// open the jar
+			if jarFile, err := os.Open(path); err == nil {
```

Comment:
You should return on `err != nil` instead to avoid the pyramid of doom.

Reply:
I'd like to fall back to decoding META-INF and then jarname, and it seems easier to do this with nesting `err == nil` then returning an error case every time

Reply:
You should refactor this pyramid of doom into a separate function that returns on `err != nil` and then call that function. This nesting seriously affects readability.